rcl_action/src/rcl_action/action_server.c

```diff
@@ -451,13 +451,28 @@ _recalculate_expire_timer(
     if (!rcl_action_goal_handle_is_active(goal_handle)) {
       ++num_inactive_goals;
 
-      rcl_action_goal_info_t goal_info;
-      ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
+      rcl_time_point_value_t goal_terminal_timestamp;
+      ret = rcl_action_goal_handle_get_goal_terminal_timestamp(
+        goal_handle, &goal_terminal_timestamp);
       if (RCL_RET_OK != ret) {
         return RCL_RET_ERROR;
       }
 
-      int64_t delta = timeout - (current_time - _goal_info_stamp_to_nanosec(&goal_info));
+      // After state of goal is updated to terminal state (success or cancel or abort),
+      // rcl_action_notify_goal_done() is called.
+      // This function is called in rcl_action_notify_goal_done().
+      // If goal_terminal_timestamp of goal is invaild, this goal is just in terminal state.
+      // So current time is set to goal_terminal_timestamp of this goal.
+      if (goal_terminal_timestamp == 0) {
+        goal_terminal_timestamp = current_time;
+        ret = rcl_action_goal_handle_set_goal_terminal_timestamp(
+          goal_handle, goal_terminal_timestamp);
+        if (RCL_RET_OK != ret) {
+          return RCL_RET_ERROR;
+        }
+      }
```

Comment:
Sorry, I didn't quite understand what you meant.   
Do you mean ?
```c++
             if (RCL_RET_OK != ret) {
                 return ret;
             }
```

Reply:
> This one should only be executed, if a goal finishes. Having it in the expire function makes the code flow harder to read, and may have unintended side effects.

Due to performance considerations, I didn't put the code into rcl_action_notify_goal_done(). If it were placed in rcl_action_notify_goal_done(), it would require looping through all the goal_handles in both rcl_action_notify_goal_done() and _recalculate_expire_timer(), and rcl_action_goal_handle_get_goal_terminal_timestamp() would be called twice.  

Of course, this does impact the readability of the code.

Reply:
Goal do no finish in high frequencies, therefore I would go for readability over performance.

Also, from my point of view, this should NOT be executed in _recalculate_expire_timer().
This code sets the timestamp of the goal termination, why would you call it in the case of the expire timer ?

Reply:
> Goals do no finish in high frequencies, therefore I would go for readability over performance.

Yes.

> Also, from my point of view, this should NOT be executed in _recalculate_expire_timer().
> This code sets the timestamp of the goal termination, why would you want to do this in the expire timer ?

I put it in _recalculate_expire_timer because I considered performance issues. I agree with what you said above, so I'll move this part of the code to rcl_action_notify_goal_done().